Add DataLocation model and update storage data relation to many-to-many

hadalin opened the pull request.

dblenkus commented on resolwe/flow/models/data.py:
```diff
@@ -430,14 +441,28 @@ def save(self, render_name=False, *args, **kwargs):  # pylint: disable=keyword-a
                 self.create_entity()
 
+    @transaction.atomic
+    def _save(self, *args, **kwargs):
```
Why have you added a separate method for this? There is already a ton of stuff in existing `save`.

hadalin replied:
This `_save` is the absolute minimum _save_. If you, for example, want to implement duplicate function, one would call `_save` inside it, because you don't need validation and whatnot, but you still need to create `DataLocation` instance.

dblenkus replied:
This function is too specific to be used in duplication function as it sets data location to data it, which is only true when new data object is created.

dblenkus replied:
I agree that this split is useful, I just wanted to say that it doesn't support all use-cases, because it sets `location_id=str(self.pk)`.
And i am in favour of a more descriptive name instead of `_save` - maybe `perform_save`, to follow DRF's practice.

hadalin replied:
>And i am in favour of a more descriptive name instead of _save - maybe perform_save, to follow DRF's practice.

That makes it a "public" function. Does it need to be decorated with `@transaction.atomic`?

dblenkus replied:
Good question :).

If a Data object is created over the API, there are at least three transactions  ([here](https://github.com/genialis/resolwe/blob/master/resolwe/flow/views/data.py#L174), [here](https://github.com/genialis/resolwe/blob/master/resolwe/flow/models/data.py#L432) and this one). And there are still lines like [this one](https://github.com/genialis/resolwe/blob/master/resolwe/flow/models/data.py#L398) that should be in s transaction, but are not always (e.g. when executor updates the output).

A good solution nay be to name the method `_perform_create` and remove the transaction.

dblenkus commented on resolwe/flow/models/data.py:
```diff
@@ -486,3 +511,13 @@ class DataDependency(models.Model):
+    data = models.OneToOneField(Data, on_delete=models.CASCADE, primary_key=True, related_name='location')
+
+    #: location identifier bound to data object's id
+    location_id = models.CharField(max_length=255)
```
You can set a lower length, e.g. 30.

dblenkus replied:
And we want to index this as it will be used in purge.

hadalin commented on resolwe/flow/tests/test_models.py:
```diff
@@ -663,23 +663,6 @@ def test_save_storage_file(self):
         self.assertEqual(data.output['json_field'], storage.pk)
         self.assertEqual(storage.json, {'foo': 'bar'})
 
-    def test_delete_data(self):
```
No, `Storage`s are deleted in purge, similarly to unreferenced data object's directory. Many data objects can point to the same storage.

dblenkus replied:
As discussed, this should be handled in a signal. If you call purge with data id after data object is deleted, the current code doesn't leave any left-overs. So we have to keep this functionality.

hadalin replied:
>If you call purge with data id after data object is deleted, the current code doesn't leave any left-overs.

Not entirely true. Data objects directories are [deleted only when purge is called without data ids](https://github.com/genialis/resolwe/blob/master/resolwe/flow/utils/purge.py#L126-L141).

My reasoning was to do the same with storages: leave it until purge without ids is called.

dblenkus replied:
> Data objects directories are deleted only when purge is called without data ids.

Not entirely true ;). You just have to add `data_purge(data_id)` in API after deleting data object and it works as described above.

hadalin commented on resolwe/flow/models/utils.py:
```diff
@@ -435,32 +435,33 @@ def add_dir_size(obj):
-                for obj in value:
-                    add_dir_size(obj)
-                    data_size += obj.get('total_size', 0)
+    if data.pk:
```
It's a bugfix and an optimization. Data should be hydrated only if the model instance exists.

dblenkus replied:
So, it has to be in a separate commit.

hadalin replied:
I don't have any numbers, but one reason for this change was because tests failed due to location not being created yet.

dblenkus replied:
It looks like tests are doing it in a wrong sequence. `output` should not be populated if `DataLocation` doesn't exist. If `output` is empty the for loop should never execute as `iterate_fields` returns an empty list.

hadalin replied:
This method calculates file and directory sizes. If data objects is not yet created, it also means is does not have a location so there's not point in continuing. I changed the if statement to `if data.location:` which makes the if statement self-explanatory.

dblenkus replied:
I agree that `data.location` is much better. But file and directory sizes are retrieved from `data.output`. So it would make even more sense to use that - but note that this condition is already (implicitly) included in the following `for`, if `data.output` is empty, the loop will never execute.

hadalin commented on resolwe/flow/utils/purge.py:
```diff
@@ -136,8 +136,8 @@ def data_purge(data_ids=None, delete=False, verbosity=0):
-                # Check if a data object with the given identifier exists.
-                if not Data.objects.filter(pk=data_id).exists():
+                # Check if any data object references given identifier.
+                if not Data.objects.filter_by_location(str(data_id)):
```
True, but that would require quite a significant refactoring. For now it's at least easier to find where we deal with location, by using location related methods. I'll add a `# TODO` comment.

dblenkus replied:
This is an absolutely crucial functionality, as it can cause a massive data loss if not done right.

hadalin replied:
No shit :)

hadalin replied:
More tests it is.

dblenkus replied:
My first comment is related to the fact that variable is named `data_id` - which is not correct. If you look few lines higher, it is folder's name, so you should rename it to `location`. It assumes that location is an integer, but this is something we can live with for now.

hadalin commented on resolwe/flow/models/data.py:
```diff
@@ -430,14 +441,28 @@ def save(self, render_name=False, *args, **kwargs):  # pylint: disable=keyword-a
                 self.save_dependencies(self.input, self.process.input_schema)  # pylint: disable=no-member
                 self.create_entity()
 
+    @transaction.atomic
```
For now yes, but in the future we could call `_save` somewhere where the code is not inside atomic block. The developer might forget to wrap it.

dblenkus replied:
I would agree with this for a public method, but a private one can/should be a lot more specific and lightweight. You have to take into account that a transaction inside inside a transaction creates another savepoint and adds a bit of overhead.

hadalin commented:
Any recommended way of accessing `location_id` from within the [executor](https://github.com/genialis/resolwe/blob/9408726bc5e609d23e99fd4afe860cacc79432db/resolwe/flow/executors/docker/run.py#L150)? Only data id is passed via the [run](https://github.com/genialis/resolwe/blob/9408726bc5e609d23e99fd4afe860cacc79432db/resolwe/flow/executors/__main__.py#L51) function. Is it ok to import `Data` model inside the executor (getting some `Apps aren't loaded yet` kind of errors when running tests).

hadalin commented:
Accessing `location_id` https://github.com/genialis/resolwe/pull/571#issuecomment-448625068 is resolved [here](https://github.com/genialis/resolwe/pull/571/commits/cf9ebf440c1de9aa5c46a6810e025fad09d786c5).

dblenkus commented:
> Any recommended way of accessing location_id from within the executor?
I think it should already be present in settings passed to executor, so you can just use it from there.

dblenkus commented on resolwe/flow/models/data.py:
```diff
@@ -486,3 +508,13 @@ class DataDependency(models.Model):
+    data = models.OneToOneField(Data, on_delete=models.CASCADE, primary_key=True, related_name='location')
+
+    #: location identifier bound to data object's id
+    location_id = models.CharField(max_length=30, db_index=True)
```
`location_id` -> `location`

hadalin replied:
Thinking about this. Even `location` doesn't sound right. Location can mean anything really. It could mean absolute "physical" path, "https" path. What about `directory_name`?

hadalin replied:
That way [this](https://github.com/genialis/resolwe/pull/571/files#diff-1adac73ebf52d60a1bde30c2fceb4610R136) can be more verbose.

dblenkus replied:
This is exactly the point here, at some point we will add S3-like storage or something similar and this model has also to cover that case. It can be `location`, `path` or something similar. `location_id` is a bad choice because it is too similar to `location.id`.

dblenkus commented on docs/contributing.rst:
```diff
@@ -98,7 +98,7 @@ To run the tests with Tox_, use::
 .. _Tox: http://tox.testrun.org/
 
-Buildling documentation
+Building documentation
```
This should go into a separate commit - together with all other typos.

dblenkus commented on docs/CHANGELOG.rst:
```diff
@@ -42,6 +50,7 @@ Changed
 - Increase Docker's memory limit for 100MB to make sure processes are not
   killed when using all available memory and tune Docker memory limits to
   avoid OOM.
+  killed when using all available memory
```
This was accidentally added.

hadalin commented:
Ready for another review round.

hadalin commented on resolwe/flow/models/data.py:
```diff
@@ -486,3 +487,28 @@ class DataDependency(models.Model):
+    """Location data of the data object."""
+
+    #: location (bound to data object's id)
+    location = models.CharField(max_length=30, db_index=True)
```
Should `location` be unique?

hadalin commented on resolwe/flow/tests/test_models.py:
```diff
@@ -155,11 +154,15 @@ def test_hydrate_file_size(self):
             'output_file': {'file': 'output.txt'}
         }
 
+        data_location = DataLocation.objects.create(location=str(data.pk))
```
The `location` does not play a significant part here as long as it's unique. It's less code that way which means less complexity.

dblenkus replied:
But it can imply a wrong thing.

hadalin commented on resolwe/flow/tests/test_purge.py:
```diff
@@ -352,78 +360,117 @@ def tearDown(self):
         # and files from the second (not completed) Data objects are unchanged.
         with patch('resolwe.flow.utils.purge.os', wraps=os) as os_mock:
-            os_mock.path.isfile = MagicMock(return_value=True)
+            # os_mock.path.isfile = MagicMock(return_value=True)
```
This returns `True` no matter what. It's better for `isfile` to return the real value. Forgot to remove the commented out code.

dblenkus replied:
I agree it is better to do it without mock if it works. But then change it all over the test.

hadalin commented on resolwe/flow/models/data.py:
```diff
@@ -486,3 +487,28 @@ class DataDependency(models.Model):
+        """Compose data runtime location path."""
+        return self.get_path(prefix=settings.FLOW_EXECUTOR['RUNTIME_DIR'], filename=filename)
+
+    def get_upload_path(self, filename=None):
```
Isn't it bound to location, here `path = os.path.join(prefix, self.location)`?

dblenkus replied:
No, location is not a part of upload path. It is assembled just from `FLOW_EXECUTOR['UPLOAD_DIR']` and `filename`.

hadalin commented on resolwe/flow/migrations/0026_add_data_location.py:
```diff
@@ -0,0 +1,49 @@
+            # After migration, FLOW_EXECUTOR['DATA_DIR'] contains directories
+            # named by DataLocation IDs.
+            data_location = DataLocation.objects.create(location='')
+            data_location.location = str(data_location.id)
```
Lol, I had this, but overwritten myself. The comment stayed `# Manually set DataLocation id to preserve data directory.`

hadalin commented on resolwe/flow/models/data.py:
```diff
@@ -486,3 +487,28 @@ class DataDependency(models.Model):
+        return path
+
+    def get_runtime_path(self, filename=None):
+        """Compose data runtime location path."""
```
It's supported in `get_path`. This method just passes the argument.

hadalin commented on resolwe/flow/utils/purge.py:
```diff
@@ -97,57 +98,52 @@ def subfiles(root):
+    is_location_referenced_by_data = location.data.exists()
+    if is_location_referenced_by_data:
+        # Perform cleanup.
+        only_ok_or_error = location.data.exclude(Q(status=Data.STATUS_DONE) | Q(status=Data.STATUS_ERROR)).count() == 0
```
This assumes there's going to be periods of no data processing. What if the server is "very active" and incessantly processing? Is `purge` called only on downtime?

dblenkus replied:
Sorry, i meant `if location.data.exclude...: return`

hadalin commented on resolwe/flow/utils/purge.py:
```diff
@@ -97,57 +98,52 @@ def subfiles(root):
-                if not Data.objects.filter(pk=data_id).exists():
-                    unreferenced_files.add(directory_path)
+    unreferenced_files = set()
+    is_location_referenced_by_data = location.data.exists()
```
The `is_location_referenced_by_data` is used in two places and longer names adds to readability. No?

dblenkus replied:
Longer, not too long :). But each variable also adds some overhead as you have to figure out where it comes from. In this case i would argue to use `location.data.exists()` in both cases, especially as it is shorter as `is_location_referenced_by_data`.

hadalin replied:
Using a variable with a meaningful name saves you from writing comments. You also pay extra attention when changing functionality because it forces you look where that variable is used.

dblenkus replied:
For me `location.data.exists()` is more meaningful than `is_location_referenced_by_data` if I only see a chunk of code.
I won't argue to remove it if you prefer it this way (it saves one db query, so it is better after all). But please find a shorter name :)

dblenkus commented on resolwe/flow/tests/test_purge.py:
```diff
@@ -474,3 +474,47 @@ def test_remove_same_location(self):
+        self.assertEqual(Storage.objects.count(), 2)
+
+        # Delete first object.
+        data_1.delete()
```
I was thinking about something like:

```
def delete(...):
    storages = list(self.storage.all())
    super().delete(...)
    for storage in storages:
        <delete if orphaned>
```

hadalin replied:
Doesn't this have a significant performance impact? Could this produce a timeout?

hadalin replied:
Does it even make sense to have storage purge then?

hadalin replied:
>Does it even make sense to have storage purge then?

Oh, it does because there's other ways for data object to be deleted (instead via `Data.delete()`).

hadalin commented:
Addressed your comments. [This](https://github.com/genialis/resolwe/pull/571#discussion_r241381269), [this](https://github.com/genialis/resolwe/pull/571#discussion_r249302083), and [this](https://github.com/genialis/resolwe/pull/571#discussion_r249440801) comments are unresolved.

dblenkus commented on resolwe/flow/managers/dispatcher.py:
```diff
@@ -404,19 +404,25 @@ def _get_per_data_dir(self, dir_base, data_id):
+            data_location.data.add(data)
 
-        output_path = self._get_per_data_dir('DATA_DIR', data_id)
+        output_path = self._get_per_data_dir('DATA_DIR', data_location.location)
```
You are right. Leave this as is. And also notice that this comment implies that you shouldn't use `DataLocation.get_*_path` in dispatcher.

hadalin commented:
- https://github.com/genialis/resolwe/pull/571#discussion_r241381269 Removed "hydrate optimization" (`if data.pk`) and fixed the [tests](https://github.com/genialis/resolwe/pull/571/files#diff-33c302481d283000e7573635e29d1c86R541).
- https://github.com/genialis/resolwe/pull/571#discussion_r249440801 Fixed [storage](https://github.com/genialis/resolwe/pull/571/files#diff-fe592d277a39630035897e67d6967343R498) [deletion](https://github.com/genialis/resolwe/pull/571/files#diff-aa73ad8a382e284ba1af2170c5e136caR440).

mstajdohar commented on resolwe/flow/managers/dispatcher.py:
```diff
@@ -404,19 +405,29 @@ def _get_per_data_dir(self, dir_base, data_id):
+        with transaction.atomic():
+            # Create a temporary random location
+            # and then override it with data location id
+            # since object have to be created first.
```
have -> has

mstajdohar commented on resolwe/flow/models/data.py:
```diff
@@ -486,3 +487,24 @@ class DataDependency(models.Model):
+    """Location data of the data object."""
+
+    #: location (bound to data object's id)
+    location = models.CharField(max_length=30, unique=True)
```
The notation `data.location.location` is a bit cumbersome. Could you name this differently? What exactly do you store in this field? Could we call it `slug` or `segment`?

dblenkus replied:
I suggest `path`, or maybe `subpath`, as in future there will be multiple storage types (local storage, s3,...) and this will be a subpath of the storage path.

hadalin replied:
`subpath` it is.

mstajdohar commented on resolwe/flow/utils/purge.py:
```diff
@@ -97,57 +96,52 @@ def subfiles(root):
 
+    :param location_id: Id of the
+        :class:`~resolwe.flow.models.DataLocation` data objects
+        reference to.
```
typo?

mstajdohar commented on resolwe/flow/utils/purge.py:
```diff
@@ -97,57 +96,52 @@ def subfiles(root):
+        intersected_files = reduce(lambda acc, cur: acc.intersection(cur), purge_files_sets)
+        unreferenced_files.update(intersected_files)
+    else:
+        # Perform emptying.
```
Consider changing the comment to Remove data directory

dblenkus commented on resolwe/flow/migrations/0026_add_data_location.py:
```diff
@@ -0,0 +1,42 @@
+    for data in Data.objects.all():
+        with transaction.atomic():
+            # Manually set DataLocation id to preserve
+            # data directory.
```
Put the comment in one line.

dblenkus commented on resolwe/flow/managers/dispatcher.py:
```diff
@@ -574,6 +585,7 @@ def override_settings(self, **kwargs):
         elif cmd == WorkerProtocol.FINISH:
             try:
                 data_id = message[WorkerProtocol.DATA_ID]
+                data_location = Data.objects.get(id=data_id).location
```
Do the query with `DataLocation` instead of `Location` as it is more efficient (uses less joins).

hadalin commented on resolwe/flow/executors/docker/prepare.py:
```diff
@@ -37,10 +37,7 @@ def resolve_data_path(self, data=None, filename=None):
-            return os.path.join(constants.DATA_ALL_VOLUME, str(data.id))
-
-        return os.path.join(constants.DATA_ALL_VOLUME, str(data.id), filename)
+        return data.location.get_path(prefix=constants.DATA_ALL_VOLUME, filename=filename)
```
Done, but it might be confusing because a few lines above you have 

```py
if not getattr(settings, 'FLOW_DOCKER_DONT_PULL', False):
```

where `settings` are `from django.conf import settings`. Might be worth adding a comment there too.

hadalin commented on resolwe/flow/executors/prepare.py:
```diff
@@ -84,10 +85,7 @@ def resolve_data_path(self, data=None, filename=None):
-            return os.path.join(settings.FLOW_EXECUTOR['DATA_DIR'], str(data.id))
-
-        return os.path.join(settings.FLOW_EXECUTOR['DATA_DIR'], str(data.id), filename)
+        return data.location.get_path(filename=filename)
```
Throughout this file Django's settings are used. It's not immediately obvious when one should and shouldn't use Django's settings directly.

hadalin replied:
I assume `from .global_settings import DJANGO_SETTINGS` must be used?

hadalin commented on resolwe/flow/managers/dispatcher.py:
```diff
@@ -495,10 +506,10 @@ def _prepare_context(self, data_id, data_dir, runtime_dir, **kwargs):
         """Copy executor sources into the destination directory.
 
-        :param data_id: The :class:`~resolwe.flow.models.Data` object id
+        :param data: The :class:`~resolwe.flow.models.Data` object
```
Ok, but putting `being` into this line makes it 73 characters long.

hadalin commented on resolwe/flow/managers/dispatcher.py:
```diff
@@ -28,7 +29,8 @@
 from resolwe.flow.execution_engines import ExecutionError
-from resolwe.flow.models import Data, DataDependency, Process
+from resolwe.flow.models import Data, DataDependency, DataLocation
+from resolwe.flow.models.process import Process
```
Lapsus.

dblenkus commented on resolwe/flow/utils/purge.py:
```diff
@@ -164,6 +164,24 @@ def _location_purge_all(delete=False, verbosity=0):
         logger.info("No data locations")
 
 
+def storage_purge(delete=False, verbosity=0):
```
It is not intuitive that `location_purge` purges single location and `storage_purge` purges all storages.
I don't see a use-case where this function would be used separately. So you can just rename it to `_storage_purge_all` and only use it in `purge_all`.

dblenkus commented on resolwe/flow/utils/purge.py:
```diff
@@ -164,6 +164,24 @@ def _location_purge_all(delete=False, verbosity=0):
+
+    if verbosity >= 1:
+        if orphaned_storages is not None:
+            if orphaned_storages:
```
Can you combine this `if` with the one above?

dblenkus commented:
And note that purge has changed on the master branch, so pay attention when rebasing.

hadalin commented on resolwe/flow/managers/dispatcher.py:
```diff
@@ -574,6 +584,7 @@ def override_settings(self, **kwargs):
         elif cmd == WorkerProtocol.FINISH:
             try:
                 data_id = message[WorkerProtocol.DATA_ID]
+                data_location = DataLocation.objects.get(data__id=data_id)
```
You mean `Data.objects.get(id=data_id).location.first()`? I think it was this way before, but we changed it.

hadalin commented on resolwe/flow/tests/test_models.py:
```diff
@@ -531,45 +537,65 @@ def test_process_isactive(self):
         super().setUp()
 
-        self.process = Process(
+        self.process = Process.objects.create(
```
Because otherwise test throws `ValueError: save() prohibited to prevent data loss due to unsaved related object 'process'.`

hadalin commented on resolwe/flow/utils/purge.py:
```diff
@@ -97,57 +96,52 @@ def subfiles(root):
+    """Print and conditionally delete files not referenced by meta data.
 
+    :param location_id: Id of the
+        :class:`~resolwe.flow.models.DataLocation` model that
```
72 is mid-word. Did it as 67.

dblenkus commented on resolwe/flow/managers/dispatcher.py:
```diff
@@ -413,19 +413,29 @@ def _get_per_data_dir(self, dir_base, data_id):
+        with transaction.atomic():
+            # Create a temporary random location and then override it with data
+            # location id since object has to be created first.
+            # TODO Find a better solution than creating a temporary location.
```
`# TODO Find a better solution, e.g. defer the database constraint.`

dblenkus commented on resolwe/flow/managers/listener.py:
```diff
@@ -485,13 +485,13 @@ def handle_finish(self, obj):
                 except ChannelFull:
                     logger.warning(
-                        "Cannot triger purge because channel is full.",
+                        "Cannot trigger purge because channel is full.",
```
This should go into the previous commit.

dblenkus commented on resolwe/flow/tests/test_models.py:
```diff
@@ -645,6 +675,10 @@ def test_save_storage_file(self):
             contributor=self.contributor,
             process=self.proc,
         )
+        data_location = DataLocation.objects.create(subpath='')
```
Is data location required for testing storage?

hadalin replied:
Yes, somewhere along the way `location.get_path` gets called.

dblenkus commented on resolwe/flow/tests/test_models.py:
```diff
@@ -766,6 +800,10 @@ def test_hydrate_input_references(self):
             },
             size=0,
         )
+        data_location = DataLocation.objects.create(subpath='')
```
You should use a non-empty path for testing that path is rendered correctly.

hadalin replied:
The path gets set one line below.
